core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4830,40 +4830,59 @@ private Optional<TableVersion> extractTableVersion(Table table, Optional<Express
-                    throw semanticException(TYPE_MISMATCH, queryPeriod, format(
-                            "Type %s invalid. Temporal pointers must be of type Timestamp, Timestamp with Time Zone, or Date.",
-                            type.getDisplayName()));
-                }
-                if (pointer == null) {
-                    throw semanticException(INVALID_ARGUMENTS, queryPeriod, "Pointer value cannot be NULL");
-                }
-                Instant pointerInstant = getInstantWithRoundUp((LongTimestampWithTimeZone) coerce(type, pointer, createTimestampWithTimeZoneType(TimestampWithTimeZoneType.MAX_PRECISION)));
-                if (!pointerInstant.isBefore(session.getStart())) {
-                    String varchar = ((Slice) coerce(type, pointer, createUnboundedVarcharType())).toStringUtf8();
-                    throw semanticException(
-                            INVALID_ARGUMENTS,
-                            queryPeriod,
-                            format("Pointer value '%s' is not in the past", varchar));
-                }
+            return Optional.of(createTableVersion(table.getQueryPeriod().get(), pointerType, versionType, evaluatedVersion));
+        }
+
+        private TableVersion createTableVersion(QueryPeriod queryPeriod, PointerType pointerType, Type objectType, Object pointer)
+        {
+            return switch (pointerType) {
+                case TEMPORAL -> createTemporalTableVersion(queryPeriod, objectType, pointer);
+                case TARGET_ID -> createTargetIdTableVersion(queryPeriod, objectType, pointer);
+            };
+        }
+
+        private TableVersion createTargetIdTableVersion(QueryPeriod queryPeriod, Type objectType, Object pointer)
+        {
+            if (pointer == null) {
+                throw semanticException(INVALID_ARGUMENTS, queryPeriod, "Pointer value cannot be NULL");
+            }
+            return new TableVersion(PointerType.TARGET_ID, objectType, pointer);
+        }
+
+        private TableVersion createTemporalTableVersion(QueryPeriod queryPeriod, Type objectType, Object pointer)
+        {
+            // Before checking if the connector supports the version type, verify that version is a valid time-based type
+            if (!(objectType instanceof TimestampWithTimeZoneType ||
+                    objectType instanceof TimestampType ||
+                    objectType instanceof DateType)) {
+                throw semanticException(TYPE_MISMATCH, queryPeriod, format(
+                        "Type %s invalid. Temporal pointers must be of type Timestamp, Timestamp with Time Zone, or Date.",
+                        objectType.getDisplayName()));
+            }
+            if (pointer == null) {
+                throw semanticException(INVALID_ARGUMENTS, queryPeriod, "Pointer value cannot be NULL");
+            }
+            Type targetType = createTimestampWithTimeZoneType(TimestampWithTimeZoneType.MAX_PRECISION);
```

Comment:
Do we want to allow the connector to still accept Dates, if it chooses so?

Do we want to allow the connector to determine the precision of the timestamptz it wants? (i guess no-one wants picos yet)

Reply:
BTW with this change, `ConnectorTableVersion` should be explicit how the temporals are carried

Reply:
> Do we want to allow the connector to still accept Dates, if it chooses so?

I'm having a hard time to figure out why would the connector prefer something else than timestamp with time zone.

> Do we want to allow the connector to determine the precision of the timestamptz it wants? (i guess no-one wants picos yet) ?

Do you imply adding a method to `ConnectorMetadata` where the connector can state the maximum precision it expects? This sounds reasonable (at the cost of an extra operation on the `ConnectorMetadata` interface).

Reply:
> > Do we want to allow the connector to still accept Dates, if it chooses so?
> 
> I'm having a hard time to figure out why would the connector prefer something else than timestamp with time zone.

not unreasonable

cc @martint 

> > Do we want to allow the connector to determine the precision of the timestamptz it wants? (i guess no-one wants picos yet) ?
> 
> Do you imply adding a method to `ConnectorMetadata` where the connector can state the maximum precision it expects? This sounds reasonable (at the cost of an extra operation on the `ConnectorMetadata` interface).

Depends on the answer to the first question (about data types / data type families)